director/redirect.go

```diff
@@ -389,8 +389,17 @@ func registerServeAd(ctx *gin.Context, sType ServerType) {
 		Type:    sType,
 	}
 
+	hasOriginAdInCache := serverAds.Has(sAd)
 	RecordAd(sAd, &ad.Namespaces)
 
+	// Start director periodic test of origin's health status if origin AD
+	// has WebURL field AND it's not already been registered
+	serverAdMutex.RLock()
+	if ad.WebURL != "" && !hasOriginAdInCache {
```

Comment:
What happens if the origin disappears from the cache, the cache gets garbage collected, then it comes up again?  As far as this logic goes, we now have a "new" origin and thus a second goroutine testing it.

Why not pass the go routine a `context.WithCancel(context.Background()), keep a cancel function in the cache, and then call the `cancel()` on invalidation?  That would remove the running goroutine when caches disappear.

Reply:
I thought the logic in `origin_monitor.go` will return and end the goroutine when origin disappears from the cache.

Reply:
Introduced a new map to keep all cancel functions as the cache key needs to be comparable while the cancelFunc is not.